src/hotspot/share/services/threadService.cpp

```diff
@@ -1115,3 +1118,474 @@ ThreadsListEnumerator::ThreadsListEnumerator(Thread* cur_thread,
+  }
+  static void set_stack_trace(oop snapshot, oop trace) {
+    snapshot->obj_field_put(_stackTrace_offset, trace);
+  }
+  static void set_locks(oop snapshot, oop locks) {
+    snapshot->obj_field_put(_locks_offset, locks);
+  }
+  static void set_blocker(oop snapshot, int type_ordinal, oop lock) {
+    snapshot->int_field_put(_blockerTypeOrdinal_offset, type_ordinal);
+    snapshot->obj_field_put(_blockerObject_offset, lock);
+  }
+  static void set_blocker_owner(oop snapshot, oop owner) {
+    snapshot->obj_field_put(_blockerOwner_offset, owner);
+  }
+};
+
+bool jdk_internal_vm_ThreadSnapshot::_inited = false;
+int jdk_internal_vm_ThreadSnapshot::_name_offset;
+int jdk_internal_vm_ThreadSnapshot::_threadStatus_offset;
+int jdk_internal_vm_ThreadSnapshot::_carrierThread_offset;
+int jdk_internal_vm_ThreadSnapshot::_stackTrace_offset;
+int jdk_internal_vm_ThreadSnapshot::_locks_offset;
+int jdk_internal_vm_ThreadSnapshot::_blockerTypeOrdinal_offset;
+int jdk_internal_vm_ThreadSnapshot::_blockerObject_offset;
+int jdk_internal_vm_ThreadSnapshot::_blockerOwner_offset;
+
+oop VMThreadSnapshot::get_thread_snapshot(jobject jthread, TRAPS) {
+  ThreadsListHandle tlh(THREAD);
+
+  ResourceMark rm(THREAD);
+  HandleMark   hm(THREAD);
+  Handle java_thread(THREAD, JNIHandles::resolve(jthread));
+
+  // wrapper to auto delete JvmtiVTMSTransitionDisabler
+  class TransitionDisabler {
+    JvmtiVTMSTransitionDisabler* _transition_disabler;
+  public:
+    TransitionDisabler(): _transition_disabler(nullptr) {}
+    ~TransitionDisabler() {
+      reset();
+    }
+    void init(jobject jthread) {
+      _transition_disabler = new (mtInternal) JvmtiVTMSTransitionDisabler(jthread);
+    }
+    void reset() {
+      if (_transition_disabler != nullptr) {
+        delete _transition_disabler;
+        _transition_disabler = nullptr;
+      }
+    }
+  } transition_disabler;
+
+  JavaThread* thread = nullptr;
+  bool is_virtual = java_lang_VirtualThread::is_instance(java_thread());
+  Handle carrier_thread;
+  if (is_virtual) {
+    // 1st need to disable mount/unmount transitions
+    transition_disabler.init(jthread);
+
+    carrier_thread = Handle(THREAD, java_lang_VirtualThread::carrier_thread(java_thread()));
+    if (carrier_thread != nullptr) {
+      thread = java_lang_Thread::thread(carrier_thread());
+    }
+  } else {
+    thread = java_lang_Thread::thread(java_thread());
+  }
+
+  // Handshake with target
+  GetThreadSnapshotClosure cl(THREAD, java_thread, thread);
+  if (thread == nullptr) {
+    // unmounted vthread, execute on the current thread
+    cl.do_thread(nullptr);
+  } else {
+    do {
+      Handshake::execute(&cl, &tlh, thread);
+    } while (cl.read_reset_retry());
+  }
+
+  // all info is collected, can enable transitions.
+  transition_disabler.reset();
+
+  // StackTrace
+  InstanceKlass* ste_klass = vmClasses::StackTraceElement_klass();
+  assert(ste_klass != nullptr, "must be loaded in 1.4+");
```

Comment:
```suggestion
  assert(ste_klass != nullptr, "must be loaded");
```
The 1.4+ is not relevant these days